optuna/integration/skopt.py

```diff
@@ -0,0 +1,242 @@
+                               "an independent sampler, not `skopt.Optimizer`.".format(
+                                   param_name, trial.number))
+
+        return self._independent_sampler.sample_independent(study, trial, param_name,
+                                                            param_distribution)
+
+
+class _Optimizer(object):
+    def __init__(self, search_space, skopt_kwargs=None):
+        # type: (Dict[str, BaseDistribution], Optional[Dict[str, Any]]) -> None
+
+        self._search_space = search_space
+
+        dimensions = []
+        for name, distribution in sorted(self._search_space.items()):
+            if isinstance(distribution, distributions.UniformDistribution):
+                high = max(distribution.low, np.nextafter(distribution.high, float('-inf')))
+                dimension = space.Real(distribution.low, high)
+            elif isinstance(distribution, distributions.LogUniformDistribution):
+                high = max(distribution.low, np.nextafter(distribution.high, float('-inf')))
+                dimension = space.Real(distribution.low, high, prior='log-uniform')
+            elif isinstance(distribution, distributions.IntUniformDistribution):
+                dimension = space.Integer(distribution.low, distribution.high)
+            elif isinstance(distribution, distributions.DiscreteUniformDistribution):
+                count = (distribution.high - distribution.low) // distribution.q
+                dimension = space.Integer(0, count)
+            elif isinstance(distribution, distributions.CategoricalDistribution):
+                dimension = space.Categorical(distribution.choices)
+            else:
+                raise NotImplementedError(
+                    "The distribution {} is not implemented.".format(distribution))
+
+            dimensions.append(dimension)
+
+        self._optimizer = skopt.Optimizer(dimensions, **skopt_kwargs)
+
+    def tell(self, study):
+        # type: (InTrialStudy) -> None
+
+        xs = []
+        ys = []
+        for trial in study.trials:
+            if trial.state != structs.TrialState.COMPLETE:
+                continue
+
+            if not self._is_compatible(trial):
+                continue
+
+            x, y = self._complete_trial_to_skopt_observation(study, trial)
+            xs.append(x)
+            ys.append(y)
+
+        self._optimizer.tell(xs, ys)
+
+    def ask(self):
+        # type: () -> Dict[str, float]
+
+        params = {}
+        param_values = self._optimizer.ask()
+        for (name, distribution), value in zip(sorted(self._search_space.items()), param_values):
+            if isinstance(distribution, distributions.DiscreteUniformDistribution):
+                value = value * distribution.q + distribution.low
+
+            params[name] = distribution.to_internal_repr(value)
+
+        return params
+
+    def _is_compatible(self, trial):
```

Comment:
Currently, the test code does not cover the case `_is_compatible(trial) == False`.
Could you add the test cases to confirm that this method is able to filter out incompatible trials?

Reply:
Added: https://github.com/pfnet/optuna/pull/399/commits/9a2519bd6683a9f3bd902ec22617097431603800